File: app/src/main/java/org/mifos/mobilebanking/ui/fragments/HelpFragment.java
package org.mifos.mobilebanking.ui.fragments;

/*
~This project is licensed under the open source MPL V2.
~See https://github.com/openMF/self-service-app/blob/master/LICENSE.md
*/

import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.BottomNavigationView;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.models.FAQ;
import org.mifos.mobilebanking.presenters.HelpPresenter;
import org.mifos.mobilebanking.ui.activities.base.BaseActivity;
import org.mifos.mobilebanking.ui.adapters.FAQAdapter;
import org.mifos.mobilebanking.ui.fragments.base.BaseFragment;
import org.mifos.mobilebanking.ui.views.HelpView;
import org.mifos.mobilebanking.utils.Constants;
import org.mifos.mobilebanking.utils.DividerItemDecoration;

import java.util.ArrayList;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

public class HelpFragment extends BaseFragment implements HelpView, BottomNavigationView.
        OnNavigationItemSelectedListener {

    @BindView(R.id.rv_faq)
    RecyclerView rvFaq;

    @BindView(R.id.bnv_help)
    BottomNavigationView bnvHelp;

    @Inject
    FAQAdapter faqAdapter;

    @Inject
    HelpPresenter presenter;

    private View rootView;
    private ArrayList<FAQ> faqArrayList;

    public static HelpFragment newInstance() {
        HelpFragment fragment = new HelpFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_help, container, false);
        setHasOptionsMenu(true);

        ((BaseActivity) getActivity()).getActivityComponent().inject(this);
        ButterKnife.bind(this, rootView);
        presenter.attachView(this);
        setToolbarTitle(getString(R.string.help));

        showUserInterface();
        if (savedInstanceState == null) {
            presenter.loadFaq();
        }

        return rootView;
    }
    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelableArrayList(Constants.HELP, new ArrayList<Parcelable>(faqArrayList));
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (savedInstanceState != null) {
            ArrayList<FAQ> faqs = savedInstanceState.getParcelableArrayList(Constants.HELP);
            showFaq(faqs);
        }
    }

    private void showUserInterface() {

        LinearLayoutManager layoutManager = new LinearLayoutManager(getActivity());
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        rvFaq.setLayoutManager(layoutManager);
        rvFaq.addItemDecoration(new DividerItemDecoration(getActivity(),
                layoutManager.getOrientation()));

        bnvHelp.setOnNavigationItemSelectedListener(this);
    }

    @Override
    public void showFaq(ArrayList<FAQ> faqArrayList) {
        faqAdapter.setFaqArrayList(faqArrayList);
        rvFaq.setAdapter(faqAdapter);
        this.faqArrayList = faqArrayList;
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_call:
                Intent intent = new Intent(Intent.ACTION_DIAL);
                intent.setData(Uri.parse("tel:" + getString(R.string.help_line_number)));
                startActivity(intent);
                break;
            case R.id.menu_email:
                Intent emailIntent = new Intent(Intent.ACTION_SENDTO);
                emailIntent.setData(Uri.parse("mailto:" + getString(R.string.contact_email)));
                emailIntent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.user_query));
                if (emailIntent.resolveActivity(getActivity().getPackageManager()) != null) {
                    startActivity(emailIntent);
                }
                break;
            case R.id.menu_locations:
                ((BaseActivity) getActivity()).replaceFragment(LocationsFragment.newInstance(),
                        true, R.id.container);
                break;
        }
        return true;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_help, menu);

        SearchManager searchManager = (SearchManager) getActivity().getSystemService(Context.
                SEARCH_SERVICE);

        SearchView searchView = (SearchView) menu.findItem(R.id.menu_search_faq).getActionView();
        searchView.setSearchableInfo(searchManager.getSearchableInfo(getActivity().
                getComponentName()));
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                faqAdapter.updateList(presenter.filterList(faqArrayList, newText));
                return false;
            }
        });
    }

    @Override
    public void showProgress() {
        showProgressBar();
    }

    @Override
    public void hideProgress() {
        hideProgressBar();
    }
}

File: app/src/main/java/org/mifos/mobile/ui/fragments/HelpFragment.kt
package org.mifos.mobile.ui.fragments

import android.app.SearchManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import com.github.therajanmaurya.sweeterror.SweetUIErrorHandler
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import org.mifos.mobile.R
import org.mifos.mobile.databinding.FragmentHelpBinding
import org.mifos.mobile.models.FAQ
import org.mifos.mobile.ui.activities.base.BaseActivity
import org.mifos.mobile.ui.adapters.FAQAdapter
import org.mifos.mobile.ui.fragments.base.BaseFragment
import org.mifos.mobile.utils.Constants
import org.mifos.mobile.utils.DividerItemDecoration
import org.mifos.mobile.utils.HelpUiState
import org.mifos.mobile.viewModels.HelpViewModel
import javax.inject.Inject

/*
~This project is licensed under the open source MPL V2.
~See https://github.com/openMF/self-service-app/blob/master/LICENSE.md
*/
@AndroidEntryPoint
class HelpFragment : BaseFragment() {
    private var _binding: FragmentHelpBinding? = null
    private val binding get() = _binding!!

    @JvmField
    @Inject
    var faqAdapter: FAQAdapter? = null

    private val viewModel: HelpViewModel by viewModels()

    private var faqArrayList: ArrayList<FAQ?>? = null
    private lateinit var sweetUIErrorHandler: SweetUIErrorHandler

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        _binding = FragmentHelpBinding.inflate(inflater, container, false)
        val rootView = binding.root
        setHasOptionsMenu(true)
        setToolbarTitle(getString(R.string.help))
        sweetUIErrorHandler = SweetUIErrorHandler(activity, rootView)
        showUserInterface()
        if (savedInstanceState == null) {
            viewModel.loadFaq(
                context?.resources?.getStringArray(R.array.faq_qs),
                context?.resources?.getStringArray(R.array.faq_ans)
            )
        }
        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewLifecycleOwner.lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.helpUiState.collect() {
                    when (it) {

                        is HelpUiState.ShowFaq -> {
                            showFaq(it.faqArrayList)
                        }

                        HelpUiState.Initial -> {}
                    }
                }
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelableArrayList(Constants.HELP, ArrayList<Parcelable?>(faqArrayList))
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState != null) {
            val faqs: ArrayList<FAQ?> =
                savedInstanceState.getParcelableArrayList(Constants.HELP) ?: arrayListOf()
            showFaq(faqs)
        }
    }

    private fun showUserInterface() {
        val layoutManager = LinearLayoutManager(activity)
        layoutManager.orientation = LinearLayoutManager.VERTICAL
        binding.rvFaq.layoutManager = layoutManager
        binding.rvFaq.addItemDecoration(
            DividerItemDecoration(
                activity,
                layoutManager.orientation,
            ),
        )
        binding.callButton.setOnClickListener {
            val intent = Intent(Intent.ACTION_DIAL)
            intent.data = Uri.parse("tel:" + getString(R.string.help_line_number))
            startActivity(intent)
        }
        binding.mailButton.setOnClickListener {
            val intent = Intent(Intent.ACTION_SENDTO).apply {
                data = Uri.parse("mailto:")
                putExtra(Intent.EXTRA_EMAIL, arrayOf(getString(R.string.contact_email)))
                putExtra(Intent.EXTRA_SUBJECT, getString(R.string.user_query))
            }
            try {
                startActivity(intent)
            } catch (e: Exception) {
                Toast.makeText(
                    requireContext(),
                    getString(R.string.no_app_to_support_action),
                    Toast.LENGTH_SHORT,
                ).show()
            }
        }
        binding.locationsButton.setOnClickListener {
            (activity as BaseActivity?)?.replaceFragment(
                LocationsFragment.newInstance(),
                true,
                R.id.container,
            )
        }
    }

    private fun showFaq(faqArrayList: ArrayList<FAQ?>?) {
        faqAdapter?.setFaqArrayList(faqArrayList)
        binding.rvFaq.adapter = faqAdapter
        this.faqArrayList = faqArrayList
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_help, menu)
        val searchManager = activity?.getSystemService(Context.SEARCH_SERVICE) as SearchManager
        val searchView = menu.findItem(R.id.menu_search_faq).actionView as SearchView
        searchView.setSearchableInfo(searchManager.getSearchableInfo(activity?.componentName))
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                val filteredFAQList = viewModel.filterList(faqArrayList, newText)
                filteredFAQList.let {
                    if (it.isNotEmpty()) {
                        sweetUIErrorHandler.hideSweetErrorLayoutUI(
                            binding.rvFaq,
                            binding.layoutError.clErrorLayout,
                        )
                        faqAdapter?.updateList(it)
                    } else {
                        showEmptyFAQUI()
                    }
                } ?: showEmptyFAQUI()
                return false
            }
        })
    }

    private fun showEmptyFAQUI() {
        sweetUIErrorHandler.showSweetEmptyUI(
            getString(R.string.questions),
            R.drawable.ic_help_black_24dp,
            binding.rvFaq,
            binding.layoutError.clErrorLayout,
        )
    }

    fun showProgress() {
        showProgressBar()
    }

    fun hideProgress() {
        hideProgressBar()
    }

    companion object {
        @JvmStatic
        fun newInstance(): HelpFragment {
            val fragment = HelpFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }

    override fun onDestroyView() {
        _binding = null
        super.onDestroyView()
    }
}
